src/Features/Core/Portable/ConvertForToForEach/AbstractConvertForToForEachCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,503 @@
+                        {
+                            return true;
+                        }
+                    }
+                }
+
+                return false;
+            }
+        }
+
+        private IEnumerable<TSymbol> TryFindMembersInThisOrBaseTypes<TSymbol>(
+            INamedTypeSymbol containingType, ITypeSymbol type, string memberName) where TSymbol : class, ISymbol
+        {
+            var methods = type.GetAccessibleMembersInThisAndBaseTypes<TSymbol>(containingType);
+            return methods.Where(m => m.Name == memberName);
+        }
+
+        private TSymbol TryFindMemberInThisOrBaseTypes<TSymbol>(
+            INamedTypeSymbol containingType, ITypeSymbol type, string memberName) where TSymbol : class, ISymbol
+        {
+            return TryFindMembersInThisOrBaseTypes<TSymbol>(containingType, type, memberName).FirstOrDefault();
+        }
+
+        private bool TryGetIterationElementType(
+            INamedTypeSymbol containingType, ITypeSymbol collectionType, 
+            INamedTypeSymbol ienumerableType, INamedTypeSymbol ienumeratorType,
+            out ITypeSymbol iterationType)
+        {
+            if (collectionType is IArrayTypeSymbol arrayType)
+            {
+                iterationType = arrayType.ElementType;
+
+                // We only support single-dimensional array iteration.
+                return arrayType.Rank == 1;
+            }
+
+            // Check in the class/struct hierarchy first.
+            var getEnumeratorMethod = TryFindMemberInThisOrBaseTypes<IMethodSymbol>(
+                containingType, collectionType, WellKnownMemberNames.GetEnumeratorMethodName);
+            if (getEnumeratorMethod != null)
+            {
+                return TryGetIterationElementTypeFromGetEnumerator(
+                    containingType, getEnumeratorMethod, ienumeratorType, out iterationType);
+            }
+
+            // couldn't find .GetEnumerator on the class/struct.  Check the interface hierarchy.
+            var instantiatedIEnumerableType = collectionType.GetAllInterfacesIncludingThis().FirstOrDefault(
+                t => Equals(t.OriginalDefinition, ienumerableType));
+
+            if (instantiatedIEnumerableType != null)
+            {
+                iterationType = instantiatedIEnumerableType.TypeArguments[0];
+                return true;
+            }
+
+            iterationType = default;
+            return false;
+        }
+
+        private bool TryGetIterationElementTypeFromGetEnumerator(
+            INamedTypeSymbol containingType, IMethodSymbol getEnumeratorMethod, 
+            INamedTypeSymbol ienumeratorType, out ITypeSymbol iterationType)
+        {
+            var getEnumeratorReturnType = getEnumeratorMethod.ReturnType;
+
+            // Check in the class/struct hierarchy first.
+            var currentProperty = TryFindMemberInThisOrBaseTypes<IPropertySymbol>(
```

Comment:
can this just use FindImplementationForInterfaceMember?

Reply:
then, you can simply do 
var property = containingtype.FindIMplemenetationForInterfaceMember(ienumeratorType.GetMember("Current")) as IPropertySymbol;
property.getter.ReturnType ?

Reply:
No.  We use this lke so:

```c#
+            var getEnumeratorMethod = TryFindMemberInThisOrBaseTypes<IMethodSymbol>(
+                containingType, collectionType, WellKnownMemberNames.GetEnumeratorMethodName);
```

Effectively, we're looking for Enumerator *pattern*.  We don't know that your code actually implements IEnumerable (and, indeed, it may not).  We still want to offer to convert to foreach even if you only implement the pattern and don't use the actual IEnumerable/IEnumerator interfaces.

Reply:
yes. for GetEnumerator yet. but not for Current. it must be IEnumerator type.

Reply:
> yes. for GetEnumerator yet. but not for Current. it must be IEnumerator type.

No, that's not a requirement.  GetEnumerator simply needs to return some type.  That type simply has to have a .Next and  .Current on it.  It does not need to implement IEnumerator.

Reply:
This is totally legal:

```c#
    public E GetEnumerator() => throw new Exception();

    public struct E
    {
        public bool MoveNext() => throw new Exception();
        public int Current { get; }
    }
```

The outer type does nto need to implement IEnumerable, and the inner type does not need to implement IEnumerator.  This is a common pattern for people who want enumeration without any possibility of boxing.

Reply:
ah, yes. I remember those trick compiler had in their code.